template: Add option to `lstrip_blocks' and fix setting`trim_blocks` inline

alextsits opened the pull request:
[lstrip_blocks_test.zip](https://github.com/ansible/ansible/files/1816249/lstrip_blocks_test.zip)

[trim_blocks_test.zip](https://github.com/ansible/ansible/files/1817420/trim_blocks_test.zip)

##### SUMMARY
Add option to set `lstrip_blocks` when using the template module to render Jinja templates. The Jinja documentation suggests that `trim_blocks` and `lstrip_blocks` is a great combination and the template module already provides an option for `trim_blocks'.

Note that although `trim_blocks` in Ansible is enabled by default since version 2.4, in order to avoid breaking things keep `lstrip_blocks` disabled by default. Maybe in a future version it could be enabled by default.

Also fix the `trim_blocks` argument to make it possible to set it to `False` inline.

This seems to address issue #10725 in a more appropriate way than the
suggested.

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
<!--- Name of the module, plugin, module or task -->
template

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.6.0
  config file = None
  configured module search path = [u'/home/alex/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/alex/git/ansible-ve/lib/python2.7/site-packages/ansible-2.6.0-py2.7.egg/ansible
  executable location = /home/alex/git/ansible-ve/bin/ansible
  python version = 2.7.14 (default, Jan  5 2018, 10:41:29) [GCC 7.2.1 20171224]

```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

Attached is a simple test. The zip file contains the playbook `lstrip_blocks_test.yml` and the template `test.j2`. The playbook contains two tasks. The first task should render the template without setting the `lstrip_blocks` option, creating the file `test_default` in the working directory.
The second task should render the template with the `lstrip_blocks` options set to `True`, creating the `test_with_lstrip_blocks` file in the working directory.

To run the playbook simply run `ansible-playbook lstrip_blocks_test.yml`.

To test the `trim_blocks` inline option, we provide another test. The zip file contains a playbook `trim_blocks_test.yml` and the template `trim_blocks.j2`. The playbook has a single task, to render the `trim_blocks.j2` template with `trim_blocks` set to `False`, creating the file `trim_blocks.templated` in the working directory. Before this PR the result is the following:
```
Hello World
Goodbye
```

After the PR the result is the following (which is the correct and expected):
```

Hello World

Goodbye
```

To run the playbook simply run `ansible-playbook trim_blocks_test.yml`.

samdoran commented:
This looks good. Please add integration tests and I'll merge it.

alextsits committed:
template: Add integration tests for `lstrip_blocks'

Signed-off-by: Alex Tsitsimpis <alextsi@arrikto.com>

alextsits commented:
Added integration tests as requested.

alextsits commented:
Please do not merge this yet. It seems that `lstip_blocks`, just like `trim_blocks` does not work as expected when passed as inline argument to the template module. Will update the PR, and will fix `trim_blocks` too.

alextsits committed:
template: Fix passing `trim_blocks' inline

Fix passing `trim_blocks' option to the template module as inline
argument. Previously passing the `trim_blocks' option inline instead of
using the YAML dictionary format resulted in it always being set to
`True', even if `trim_blocks=False' was used.

Signed-off-by: Alex Tsitsimpis <alextsi@arrikto.com>

alextsits committed:
template: Add option to `lstrip_blocks'

Add option to set `lstrip_blocks' when using the template module to
render Jinja templates. The Jinja documentation suggests that
`trim_blocks' and `lstrip_blocks' is a great combination and the
template module already provides an option for `trim_blocks'.

Note that although `trim_blocks' in Ansible is enabled by default since
version 2.4, in order to avoid breaking things keep `lstrip_blocks'
disabled by default. Maybe in a future version it could be enabled by
default.

This seems to address issue #10725 in a more appropriate way than the
suggested.

Signed-off-by: Alex Tsitsimpis <alextsi@arrikto.com>

alextsits committed:
template: Add integration tests for `trim_blocks'

Signed-off-by: Alex Tsitsimpis <alextsi@arrikto.com>

alextsits commented:
Fixed `trim_blocks` argument, added appropriate integration tests for it too.
Also updated the PR description and added an attachment to easily reproduce the `trim_blocks` bug. I'm done with changes, ready to merge.

alextsits commented:
It seems to me that tests fail on centos6 because the `lstrip_blocks` option is available from Jinja2 version 2.7 and later, and centos6 has an older version installed. If this is true, what is the correct way to deal with this? Should I skip `lstrip_blocks` tests if Jinja2 version in <2.7?

samdoran commented:
Thanks for adding tests!

Raise an `AnsibleError` exception if `lstrip_blocks` is `True` but the Jinja2 version is < 2.7. If you silently skip it, the generated file could be different than what the playbook author intended, potentially creating a misconfigured file.

Also, please state in the option description that it requires Jinja2 >= 2.7.

samdoran commented on test/integration/targets/template/tasks/main.yml:
```diff
@@ -120,6 +120,74 @@
+# VERIFY trim_blocks
+
+- name: Render a template with "trim_blocks" set to False
+  template: src=trim_blocks.j2 dest={{output_dir}}/trim_blocks_false.templated trim_blocks=False
```
Please use multi-line YAML rather than `k=v` syntax. I know the rest of the tasks use `k=v` syntax, but they need to be changed at some point.

alextsits replied:
Noted, it's just that my commit for `trim_blocks` fixes exactly that, that the `k=v` syntax for `trim_blocks` was not working. I will gladly change to multi-line syntax though if you insist.

samdoran commented on test/integration/targets/template/tasks/main.yml:
```diff
@@ -120,6 +120,74 @@
+  copy: src=trim_blocks_false.expected dest={{output_dir}}/trim_blocks_false.expected
+
+- name: compare templated file to known good trim_blocks_false
+  shell: diff -u {{output_dir}}/trim_blocks_false.templated {{output_dir}}/trim_blocks_false.expected
```
Rather than a `shell` task, you can use the `stat` module and `assert` that the `checksum` matches. See [lineinfile tests](https://github.com/ansible/ansible/blob/devel/test/integration/targets/lineinfile/tasks/main.yml#L58) for examples.

alextsits committed:
template: Check Jinja2 support for `lstrip_blocks'

Since the `lstrip_blocks' option was added in Jinja2 version 2.7, raise
an exception when `lstrip_blocks' is set but Jinja2 does not support it.
Check support for `lstrip_blocks' option by checking `jinja2.defaults'
for `LSTRIP_BLOCKS' and do not use `jinja2.__version__' because the
latter is set to `unknown' in some cases, perhaps due to bug in
`pkg_resources' in Python 2.6.6.

Also update option description to state that Jinja2 version >=2.7 is
required.

Signed-off-by: Alex Tsitsimpis <alextsi@arrikto.com>

alextsits commented:
Added a new commit to check for Jinja2 version.
Also updated the commits that add integration tests to implement the requested changes.

Note that while assertion in tests is now done with checksum, the checksum is not hardcoded in tests. The "expected" files still remain, in case anything goes wrong to know what the expected output actually is.

It also seems to me that github did not get the commits in the right order? Possibly due to overwriting and force-pushing? The commits in the branch should be in the correct order nevertheless.

Ready for review.

alextsits commented:
I can't seem to pass the centos6 tests. I added a little debugging, printing the jinja2 version detected in the system. `jinja2.__version__` evaluates to `unknown` in centos6 image in the template module, but not in the playbook task. I don't know why this happens.
I found [this](https://centos.pkgs.org/6/centos-x86_64/python-jinja2-2.2.1-3.el6.x86_64.rpm.html). The top entry in the changelog is relevant. Perhaps centos6 image needs update?
Waiting for feedback.

samdoran commented:
I'm digging into this and see it's block of code in Jinja2 2.2.1 `__init__`:

```python
try:
    __version__ = __import__('pkg_resources') \
        .get_distribution('Jinja2').version
except:
    __version__ = 'unknown'
```

This tries to import `pkg_resources` and inside `pkg_resources` the line `from . import py31compat` throws an ImportError. That causes `__version__` to be set to 'unknown'.

samdoran commented:
Since this looks like a bug in `pkg_resources` in Python 2.6.6, let's take a different approach. Rather than checking `__version__` since it can't be used reliably, check `jinja2.defaults` for `LSTRIP_BLOCKS`.

```patch
diff --git a/lib/ansible/plugins/action/template.py b/lib/ansible/plugins/action/template.py
index 8713773829..39c1e9f3b3 100644
--- a/lib/ansible/plugins/action/template.py
+++ b/lib/ansible/plugins/action/template.py
@@ -18,12 +18,9 @@ from __future__ import (absolute_import, division, print_function)
 __metaclass__ = type
 
 import os
-import sys
-import jinja2
 import shutil
 import tempfile
 
-from distutils.version import LooseVersion
 
 from ansible import constants as C
 from ansible.errors import AnsibleError, AnsibleFileNotFound, AnsibleAction, AnsibleActionFail
@@ -63,14 +60,14 @@ class ActionModule(ActionBase):
         # Option `lstrip_blocks' was added in Jinja2 version 2.7.
         if lstrip_blocks:
             try:
-                jinja2_version = LooseVersion(jinja2.__version__)
+                import jinja2.defaults
+            except ImportError:
+                raise AnsibleError('Unable to import Jinja2 defaults for determing Jinja2 features.')
+
+            try:
+                jinja2.defaults.LSTRIP_BLOCKS
             except AttributeError:
-                raise AnsibleError("Could not determine Jinja2 version")
-            sys.stderr.write("JINJA2_VERSION: %s\n" % jinja2.__version__)
-            if jinja2_version.vstring == "unknown":
-                raise AnsibleError("Could not determine Jinja2 version")
-            if jinja2_version < LooseVersion("2.7"):
-                raise AnsibleError("Option `lstrip_blocks' is available for Jinja2 versions >=2.7")
+                raise AnsibleError("Option `lstrip_blocks' is only available in Jinja2 versions >=2.7")
 
         wrong_sequences = ["\\n", "\\r", "\\r\\n"]
         allowed_sequences = ["\n", "\r", "\r\n"]
```

You can run tests locally with `ansible-test integration --docker centos6 template`

alextsits commented:
@samdolan seems good to me, I will apply the diff.

About running tests locally I know the command, and I have been testing successfully for `ubuntu1404` and `ubuntu1604` so far, but for some reason `centos6` fails with `ERROR: Command "docker exec <container hash> /bin/bash /root/docker.sh" returned exit status 137.` I do not have this problem with `centos7`. I'm on Arch Linux. I haven't found the time to investigate that more yet, will do at some point though.

mattclay commented:
CI failure in integration tests: https://app.shippable.com/github/ansible/ansible/runs/58672/40/tests

> test/integration/targets/template/tasks/main.yml:204 / [testhost] testhost: template : Verify templated lstrip_blocks_true matches known good using checksum that=[u'lstrip_blocks_true_result.checksum == lstrip_blocks_true_good.stat.checksum']

> failure: The conditional check 'lstrip_blocks_true_result.checksum == lstrip_blocks_true_good.stat.checksum' failed. The error was: error while evaluating conditional (lstrip_blocks_true_result.checksum == lstrip_blocks_true_good.stat.checksum): 'dict' object has no attribute 'checksum'

```
{
"msg": "The conditional check 'lstrip_blocks_true_result.checksum == lstrip_blocks_true_good.stat.checksum' failed. The error was: error while evaluating conditional (lstrip_blocks_true_result.checksum == lstrip_blocks_true_good.stat.checksum): 'dict' object has no attribute 'checksum'"
}
```

alextsits commented:
This is why the integration tests fail:

The command inside the `template.py` module that results in an exception is the following:
```
try:
    jinja2.defaults.LSTRIP_BLOCKS
except AttributeError:
    raise AnsibleError("Option `lstrip_blocks' is only available in Jinja2 versions >=2.7")
```
This is an expected behavior.

The same command in the `main.yml` succeeds, resulting in skipping other tests than the expected:
```
- name: Check support for lstrip_blocks in Jinja2
  shell: python -c 'import jinja2; jinja2.defaults.LSTRIP_BLOCKS'
  register: lstrip_block_support
  ignore_errors: True
```

I had also noticed that previously when we attempted to access the `jinja2.__version__` attribute, in the `template.py` module the value `unknown` was returned, while in `main.yml` the attribute had the value `2.10`.

I don't know why the YAML and the Python module use different environments, I haven't checked the exact setup used to run the tests, yet.

jsumners commented:
I really wish this could be set to `true` by default on a global basis. Ideally I'd be able to put the setting in my `ansible.cfg` and forget about it.

mattclay commented on test/integration/targets/template/tasks/main.yml:
```diff
@@ -120,6 +120,93 @@
+# VERIFY lstrip_blocks
+
+- name: Check support for lstrip_blocks in Jinja2
+  shell: python -c 'import jinja2; jinja2.defaults.LSTRIP_BLOCKS'
```
Use `{{ ansible_python_interpreter }}` instead of `python`.

alextsits replied:
Thanks, this indeed did the trick. I just did `grep -R jinja2.__version__` in `test` and saw another `shell:` module using plain `python -c`. My bad.

mattclay replied:
Sometimes our existing tests do things that they shouldn't, such as hard-coding references to `python` instead of using `{{ ansible_python_interpreter }}`.

samdoran replied:
The reason this fixes the test failure is because there is another integration test, `template_jinja2_latest`, which sets up a virtualenv with a newer version of Jinja2. During that test, the shell task calling `python` executes to the Python version inside the virtualenv, which passes the Jinja2 version check. But the Python being used by Ansible is `/usr/bin/python`, which would fail the Jinja2 version test and doesn't support `lstrip_blocks`.

This causes the "Render a template with "lstrip_blocks" set to True" task to run but fail when in should be skipped, making `checksum` undefined in `lstrip_blocks_true_result`.
